nntrainer/graph/network_graph.cpp

```diff
@@ -313,6 +313,12 @@ int NetworkGraph::checkInitializedGraph() {
     if (lnode->getTrainable() ||
         must_support_backwarding.find(lnode->getName()) !=
           must_support_backwarding.end()) {
+      lnode->needsCalcGradient(true);
+#ifdef ENABLE_TEST
```

Comment:
This test flag is likely to be broken. Could we stick to less optimized version, and later optimize + fix test along instead?

Reply:
In general, we want to run optimized version where we want to skip running the nodes for backwarding where its not needed, and so not allocate memory for it either.
With ENABLE_TEST, we disable such optimizations. The issue is not the memory optimization here but performance as calcDerivative runs for nodes which ideally should not.

Reply:
I agree, my concern is that by we have `#ifndef ENABLE_TEST` part is never being tested, so from #1664 there is symptoms of a bug (android build is enabling this flag to true for now).

So I am suggesting to comment this out, remove #ifdef flag and make some changes to the test while switching to the optimized version.

Reply:
> performance as calcDerivative runs for nodes which ideally should not.

I wonder which nodes are? You mean the feature extract cases which only support forwarding? If it is, then it's ok.
But other than that... None of the cases comes to my mind.

Reply:
If the first few layers are non-trainable but support backward (for example, input layer followed by multi-out layer), the multi-out layer's calcDerivative should not be called. 
This patch ensures that multi-out layer's calcDerivative is not called, and the corresponding memory is also not allocated.

Reply:
There are only three locations where ENABLE_TEST is being used (2 of which are from this PR itself). I will check the issue from 1664 and see what is the problem, but ideally, we need to run some tests with ENABLE_TEST as well.

Reply:
Marking resolves #1664 as this patch solves that issue.